File: backend/app/shared/privileges.py
from typing import Literal
from fastapi import Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.dependencies import get_db, get_current_user

PrivilegeKey = Literal[
    "admin.settings",
    "admin.excel_templates",
    "admin.notifications",
    "admin.role_privileges",
    "users.manage",
    "departments.manage",
    "master_data.manage",
    # future phases
    "project.manage",
    "notebook.manage",
    "experiment.manage",
    "atr.manage",
]

PRIVILEGE_CATALOG: list[str] = [
    "admin.settings",
    "admin.excel_templates",
    "admin.notifications",
    "admin.role_privileges",
    "users.manage",
    "departments.manage",
    "master_data.manage",
    "project.manage",
    "notebook.manage",
    "experiment.manage",
    "atr.manage",
]

DEFAULT_GRANTS: dict[str, list[str]] = {
    "master_data.manage": ["HOD"],
}

# There is no dedicated "QA" role anymore — a HOD in the QA department carries
# the same full-admin authority the old QA role used to bypass every check with.
SUPERADMIN_ROLE_CODE = "HOD"
SUPERADMIN_DEPARTMENT_CODE = "QA"

# The Administration module (Users & Roles, Role Privileges, Settings, Master
# Data) — every `require_privilege(...)` check below belongs to it — is only
# usable by QA/QC department users, regardless of role or any grant in the
# RolePrivilege table. This is a hard precondition, checked before anything else.
ADMIN_MODULE_DEPARTMENT_CODES = {"QA", "QC"}


def resolve_user_privileges(db: Session, user) -> set[str]:
    from app.models.admin import RolePrivilege

    rows = (
        db.query(RolePrivilege)
        .filter(
            RolePrivilege.role_id == user.role_id,
            RolePrivilege.is_granted.is_(True),
            RolePrivilege.department_id.is_(None),
        )
        .all()
    )
    if rows:
        return {r.privilege_key for r in rows}
    # Fall back to compiled DEFAULT_GRANTS
    return {key for key, codes in DEFAULT_GRANTS.items() if user.role.code in codes}


def user_has_privilege(db: Session, user, key: str) -> bool:
    if not user.department or user.department.code not in ADMIN_MODULE_DEPARTMENT_CODES:
        return False
    if user.role.code == SUPERADMIN_ROLE_CODE and user.department.code == SUPERADMIN_DEPARTMENT_CODE:
        return True
    return key in resolve_user_privileges(db, user)


def require_privilege(key: str):
    """FastAPI dependency factory — raises 403 when the caller lacks the privilege."""

    def _dep(
        db: Session = Depends(get_db),
        current_user=Depends(get_current_user),
    ):
        if not user_has_privilege(db, current_user, key):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail=f"Privilege '{key}' required.",
            )
        return current_user

    return Depends(_dep)


# ── Project / Notebook / Experiment creation gate ───────────────────────────────
# Only HOD and Team Lead may create Projects, Notebooks, or Experiments. Chemists
# and Analysts work exclusively within experiments already created and assigned
# to them (see ASSIGNMENT_RESTRICTED_ROLES below).
CREATOR_ROLES = {"HOD", "TL"}


def require_creator_role():
    """FastAPI dependency factory — only HOD/TL may pass."""

    def _dep(current_user=Depends(get_current_user)):
        if current_user.role.code not in CREATOR_ROLES:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Only HOD or Team Lead can perform this action.",
            )
        return current_user

    return Depends(_dep)


# ── Notebook/experiment assignment gate ─────────────────────────────────────────
# Chemists and Analysts may only view/work on notebooks (and the experiments that
# live inside them) they have been explicitly assigned to via NotebookPermission.
# HOD/TL keep their existing (project-membership-based) access untouched.
ASSIGNMENT_RESTRICTED_ROLES = {"CHEM", "ANALYST"}


def assert_notebook_access(db: Session, user, notebook) -> None:
    """Raise 403 if `user` is assignment-restricted and not assigned to `notebook`."""
    if user.role.code not in ASSIGNMENT_RESTRICTED_ROLES:
        return
    from app.models.notebook import NotebookPermission

    assigned = (
        db.query(NotebookPermission)
        .filter(
            NotebookPermission.notebook_id == notebook.id,
            NotebookPermission.user_id == user.id,
            NotebookPermission.can_view.is_(True),
        )
        .first()
    )
    if not assigned:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You are not assigned to this notebook.",
        )

File: backend-node/src/shared/privileges.ts
import { User } from '../models/User.model'
import { Role } from '../models/Role.model'
import { Department } from '../models/Department.model'
import { RolePrivilege } from '../models/RolePrivilege.model'
import { ForbiddenError } from '../utils/errors'
import { Request, Response, NextFunction } from 'express'

export const PRIVILEGE_CATALOG: Record<string, string> = {
  'admin.settings': 'Manage global settings',
  'admin.excel_templates': 'Manage Excel templates',
  'admin.notifications': 'Manage notifications',
  'admin.role_privileges': 'Manage role privileges',
  'users.manage': 'Manage users',
  'departments.manage': 'Manage departments',
  'labs.manage': 'Manage labs',
  'master_data.manage': 'Manage master data',
  'calc_templates.manage': 'Manage calc templates',
  'project.manage': 'Manage projects',
  'notebook.manage': 'Manage notebooks',
  'experiment.manage': 'Manage experiments',
  'atr.manage': 'Manage ATR forms',
  'ard.manage': 'Manage ARD module',
}

// QA/QC/AD/IT departments always have access; HOD in QA too (legacy bypass)
const PRIVILEGED_DEPT_CODES = ['QA', 'QC', 'AD', 'IT']

export async function resolveAdminPrivileges(user: User): Promise<string[]> {
  const role = user.role as Role | undefined || await Role.findByPk(user.roleId || '')
  if (!role) return []

  if (role.code === 'SUPER_ADMIN' || role.code === 'DQA') {
    return Object.keys(PRIVILEGE_CATALOG)
  }

  const dept = user.department as Department | undefined ||
    (user.departmentId ? await Department.findByPk(user.departmentId) : null)

  if (!dept || !PRIVILEGED_DEPT_CODES.includes(dept.code)) return []

  if (role.code === 'HOD' && dept.code === 'QA') {
    return Object.keys(PRIVILEGE_CATALOG)
  }

  const grants = await RolePrivilege.findAll({
    where: { roleId: user.roleId || '', isGranted: true },
  })
  return grants.map(g => g.privilegeKey)
}

export async function userHasPrivilege(user: User, key: string): Promise<boolean> {
  const role = user.role as Role | undefined || await Role.findByPk(user.roleId || '')
  if (!role) return false

  // Super admin and DQA always have all privileges
  if (role.code === 'SUPER_ADMIN' || role.code === 'DQA') return true

  // Must be in privileged department
  const dept = user.department as Department | undefined ||
    (user.departmentId ? await Department.findByPk(user.departmentId) : null)

  if (!dept || !PRIVILEGED_DEPT_CODES.includes(dept.code)) return false

  // Legacy: HOD in QA has all privileges
  if (role.code === 'HOD' && dept.code === 'QA') return true

  // Check role_privileges table
  const grant = await RolePrivilege.findOne({
    where: { roleId: user.roleId || '', privilegeKey: key, isGranted: true },
  })
  return !!grant
}

export function requirePrivilege(key: string) {
  return async (req: Request, _res: Response, next: NextFunction) => {
    try {
      const user = req.user!
      const has = await userHasPrivilege(user, key)
      if (!has) throw new ForbiddenError()
      next()
    } catch (err) {
      next(err)
    }
  }
}

// Only HOD, TL, SUPER_ADMIN can create projects/notebooks/experiments
export const CREATOR_ROLES = ['HOD', 'TL', 'SUPER_ADMIN']

export async function requireCreatorRole(user: User): Promise<void> {
  const role = user.role as Role | undefined || await Role.findByPk(user.roleId || '')
  if (!role || !CREATOR_ROLES.includes(role.code)) {
    throw new ForbiddenError('Only HOD, TL, or Admin can perform this action.')
  }
  // QA dept users are view-only
  const dept = user.department as Department | undefined ||
    (user.departmentId ? await Department.findByPk(user.departmentId) : null)
  if (dept?.code === 'QA') {
    throw new ForbiddenError('QA department users are view-only for this resource.')
  }
}
